Wait for current tasks to finish before shutdown in BackbeatConsumer

welansari opened the pull request:
To avoid processing Kafka messages multiple times by different consumers, the BackbeatConsumer needs to wait for the current jobs to finish before committing the updated offsets.

This has to be done twice, once when Kafka rebalances and the second on shutdown.

This PR addresses the shutdown part.

Issue: BB-455

Commit:
Wait for current tasks to finish before shutdown in BackbeatConsumer

To avoid processing Kafka messages multiple times by different consumers,
BackbeatConsumer needs to wait for the current jobs to finish before
committing the updated offsets.

Issue: BB-455

bert-e commented:
# Hello kerkesni,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://bert-e.scality.net/scality/backbeat/doc/user).





*Status report is not available.*

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue BB-455 contains:


* *None*



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.6.29`

* `8.7.0`


Please check the `Fix Version/s` of BB-455, or the target
branch of this pull request.

bert-e commented:
# Request integration branches



Waiting for integration branch creation to be requested by the user.

To request integration branches, please comment on this pull request with the following command:

```
/create_integration_branches
```

Alternatively, the `/approve` and `/create_pull_requests` commands will automatically
create the integration branches.

francoisferrand commented on lib/BackbeatConsumer.js:
```diff
@@ -778,6 +779,17 @@ class BackbeatConsumer extends EventEmitter {
+                this._isShuttingDown = true;
+                // waiting for all current jobs to end before exiting
+                if (this._processingQueue && !this._processingQueue.idle()) {
+                    this._processingQueue.drain = next;
```
Should we not _call_ drain?
```suggestion
                    this._processingQueue.drain(next);
```

welansari replied:
Welp... it seems we have multiple dependencies of different versions of the same library `async`.
In version 2.x of `async`, the drain looks like [this](https://github.com/caolan/async/blob/c6bdaca4f9175c14fc655d3783c6af6a883e6514/lib/queue.js#L86-L89):
```
q.drain = function() {
    console.log('all items have been processed');
};
```
And in 3.x, like [this](https://github.com/caolan/async/blob/72de392b6dffc466419ae7cab419d3b120d2459d/lib/queue.js#L78-L81)
```
q.drain(() => {
    console.log('all done')
})
```
I'm not sure yet how the choice of version is made, but the 2.x works locally and on the CI, and the 3.x implementation works on tests in an actual Artesca lab :/

I pushed the code just to test if it passes in the CI, but i stil have to figure out the dependency issue before merging anything.

welansari replied:
seems something was wrong with the image i was testing with. The code does indeed only use version 2.x

francoisferrand commented on lib/BackbeatConsumer.js:
```diff
@@ -783,9 +784,31 @@ class BackbeatConsumer extends EventEmitter {
+                        });
+                        next();
+                    };
+                    return;
```
nit: not needed (there is no other return here)

welansari replied:
if i don't return here `process.nextTick(next);` will be called and we won't wait for the drain

francoisferrand commented on lib/BackbeatConsumer.js:
```diff
@@ -783,9 +784,31 @@ class BackbeatConsumer extends EventEmitter {
+                this._isShuttingDown = true;
+                // waiting for all current jobs to end before exiting
+                if (this._processingQueue && !this._processingQueue.idle()) {
+                    this._processingQueue.drain = () => {
```
* to be on the safe side, we should reset the drain callback once it is called
* what happens if there is a rebalance during shutdown? we may have an issue, since we overwrite the callback there as well... :-/

welansari replied:
Since we block the polls when shutting down, the rebalance callback will not be called.

If a rebalance was already in progress, the drain function will be overwritten, and as soon as the jobs finish the consumer will commit and disconnect.

The commit will work as the `unassign` of the rebalance will not be called because of the overwrite.

welansari replied:
used the rebalance callback to handle all the logic, we shouldn't have these problems anymore

francoisferrand commented on lib/BackbeatConsumer.js:
```diff
@@ -797,7 +820,7 @@ class BackbeatConsumer extends EventEmitter {
                     this._zookeeper.close();
                 }
-                if (this._consumer) {
+                if (this._consumer && this._consumer.isConnected()) {
```
when is this happening?

welansari replied:
This is linked to your PR, when a rebalance times out and the consumer gets disconnected we don't want in that case to wait for the disconnect event.

francoisferrand commented on lib/BackbeatConsumer.js:
```diff
@@ -381,8 +382,8 @@ class BackbeatConsumer extends EventEmitter {
         const nNewConsumeRequests = this._getAvailableSlotsInPipeline();
-        if (nNewConsumeRequests === 0) {
-            // processing pipeline is 100% busy, do not attempt to consume
+        if (nNewConsumeRequests === 0 || this._isShuttingDown) {
```
should we use a flag (as you did), or `pause()` the consumer?
the risk would be that we fail to poll, and get kicked out of the group without waiting for us to commit...

welansari replied:
i don't think having the consumer getting kicked is an issue here, the only case where the `close()` function gets called is when the pods shutdown.

Using the pause function already implemented is probably a bad idea as it unsubscribes from the topic, so the consumer won't be able to commit.

Using the Kafka consumer's pause function is going to add more complexity that is not really needed, as we have to track the paused topic/partitions and update them in case of an assignment from a rebalance, ...

francoisferrand replied:
I was talking about consumer.pause()

francoisferrand replied:
> as we have to track the paused topic/partitions and update them in case of an assignment from a rebalance

Do we? This is shutdown, so we should not take part in the assign...

Btw, maybe we should first unsubscribe: then wait for revoke to complete (eg drain, commi, unassign) : then disconnect and quit...

welansari replied:
good idea, we should use the logic already present in the rebalance callback to wait for jobs and commit the offsets

francoisferrand commented on lib/BackbeatConsumer.js:
```diff
@@ -923,10 +918,21 @@ class BackbeatConsumer extends EventEmitter {
+                    // waiting for current jobs to complete as well as commit
+                    // the latest offsets
+                    this._waitForUnassign = true;
+                    const interval = setInterval(() => {
```
why not "emit" a signal when unassigned, and simply waiting for that signal here?
(or other way to explicitely trigger the continuation instead of a timer)

francoisferrand commented on lib/BackbeatConsumer.js:
```diff
@@ -923,10 +918,21 @@ class BackbeatConsumer extends EventEmitter {
+                            return;
+                        }
+                        clearInterval(interval);
+                        this._consumer.disconnect();
```
if we unassign due to timeout, we may be disconnect already here?
(and that function may thus fail?)

francoisferrand commented on lib/BackbeatConsumer.js:
```diff
@@ -923,10 +921,20 @@ class BackbeatConsumer extends EventEmitter {
+                    // waiting for current jobs to complete as well as commit
+                    // the latest offsets
+                    this.once('unassign', status => {
+                        if (status === UNASSIGN_STATUS.TIMEOUT) {
```
when this happens, can we/should we wait for `disconnected` ?
disconnect() will have been called, not sure if we will get the event...

(can we write a test case for close() with a timeout on the rebalance ?)

welansari replied:
I added a test case for this.
If the event already got emitted before adding the event handler then we will never receive the event yes.

francoisferrand commented on lib/BackbeatConsumer.js:
```diff
@@ -587,22 +592,24 @@ class BackbeatConsumer extends EventEmitter {
             }
 
-            this._processingQueue.drain = () => unassign('drained');
+            this._processingQueue.drain = () => unassign(UNASSIGN_STATUS.DRAINED);
```
shouldn't we stick to the 'empty' callback?
the queue should never drain (or fill) unless we go through this revoke-ation process anyway, so it seems safer to not emit the signal...

francoisferrand commented on lib/BackbeatConsumer.js:
```diff
@@ -902,17 +909,8 @@ class BackbeatConsumer extends EventEmitter {
-                }
                 if (this._kafkaBacklogMetricsConfig) {
                     // publish offsets to zookeeper
                     return this._publishOffsetsCron(() => next());
```
publishing offsets must be done _after_ the unsusbribe is complete : so that it comes after the `commit`...

(but possibly before we un-assign, if it changes anything...)
(and definitely before the timeout and associated disconnect...)

francoisferrand replied:
maybe an easy wait to get around this is to introduce a flag to skip `disconnect` on timeout during shutdown: this way we can perform the `unsubscribe`, and when unassign'ed (and commited) we can do the offset publishing..

francoisferrand replied:
actually, maybe publishing to zookeeper should always be done?

francoisferrand commented on lib/BackbeatConsumer.js:
```diff
@@ -922,14 +920,28 @@ class BackbeatConsumer extends EventEmitter {
+                if (status !== UNASSIGN_STATUS.TIMEOUT) {
+                    this._consumer.disconnect();
                 }
+                this._consumer.once('disconnected', () => next());
```
to be on the safe side, we should check if we are already disconnected.
(`consumer.isConnected()` only looks at the state of the "javascript" consumer, not the underlying librdkafka client ; and that state is updated on `disconnected` event)

and looking at the same code, it is safe to call disconnect() while disconnecting.

so we may simple do this:
```
if (this._consumer.isConnected()) {
  this._consumer.disconnect();
  this._consumer.once('disconnected', () => next());
} else {
  next();
}
```

francoisferrand committed:
Publish offsets before unassign

This is the only place where we can do it (esp. on shutdown), since it
calls `consumer.position()`, with no param, and thus updates the position
of assigned partitions.

This however has an unfortunate effect: it puts Zookeeper in the path,
as the unassign with be delayed until we have retrieved the offsets
and written them to zookeeper.

This should however be fine in the context of a kafka rebalance, which
takes seconds or more.

Issue: BB-455

Commit:
bump version 8.6.29

Issue: BB-455

welansari commented:
/approve

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `bugfix/BB-455-clean-shutdown-backbeat-consumer` into
`development/8.6`
* [w/8.7/bugfix/BB-455-clean-shutdown-backbeat-consumer](https://github.com/scality/backbeat/compare/development/8.7...w/8.7/bugfix/BB-455-clean-shutdown-backbeat-consumer) will be merged into `development/8.7` 



The following branches will **NOT** be impacted:

* `development/7.10`
* `development/7.4`
* `development/7.70`
* `development/8.5`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```







*The following options are set:* **approve**

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/8.6`

* :heavy_check_mark: `development/8.7`



The following branches will **NOT** be impacted:

* `development/7.10`
* `development/7.4`
* `development/7.70`
* `development/8.5`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/8.6`

* :heavy_check_mark: `development/8.7`



The following branches have **NOT** changed:

* `development/7.10`
* `development/7.4`
* `development/7.70`
* `development/8.5`



Please check the status of the associated issue BB-455.

Goodbye kerkesni.
